### PR TITLE
refactor: remove assertions in `force` plot, test verify_valid_cmap

### PR DESCRIPTION
**Pull request by thatlittleboy**
## Overview

This PR is part of a larger effort of refactoring `_force.py` and `_force_matplotlib.py`. This PR covers some of the smaller / non-functional changes, so we can focus on the core functionalities in the following PR.

Description of the changes proposed in this pull request:

* Removed most `assert` statements for enforcing logic, as it is not good practice. The only exception is the use of assert to test for `IPython` to be installed. (No strong reason for this, if there is a more suitable Exception type suggested, I can change it).
* Added a set of tests for `verify_valid_cmap()` function (defined inside `_force.py`). To ensure no errors are introduced in the refactoring.
* Add docstrings for some functions in `_force.py`.
* Re-use an existing `getjs()` function to read the js bundle (DRY principle).
* Added docstring for `initjs()`, just a quality-of-life improvement.

## Checklist

- [X] All [pre-commit checks](https://pre-commit.com/#install) pass.
- [X] Unit tests added (if fixing a bug or adding a new feature)

### CODE REVIEW

File: shap/plots/_force.py [hunk 1]
```diff
@@ -90,15 +117,19 @@ def force(base_value, shap_values=None, features=None, feature_names=None, out_n
...
-                            "for multi-output models try " \
-                            "shap.force_plot(explainer.expected_value[0], shap_values[0]).")
-
+            emsg = (
+                "In v0.20, force plot now requires the base value as the first parameter! "
+                "Try shap.plots.force(explainer.expected_value, shap_values) or "
+                "for multi-output models try "
+                "shap.plots.force(explainer.expected_value[0], shap_values[0])."
+            )
+            raise Exception(emsg)
```
Comment (connortann) [inline shap/plots/_force.py]: Would a TypeError be more appropriate here, and in other instances below?

File: tests/plots/test_force.py [hunk 2]
```diff
@@ -2,9 +2,41 @@
...
+    [
+        ("coolwarm", None),
+        (["#000000", "#ffffff"], None),
+        (777, "Plot color map must be string or list!"),
+        ([], "Color map must be at least two colors"),
+        (["#8834BB"], "Color map must be at least two colors"),
+        (["#883488", "#Gg8888"], r"Invalid color .+ found in cmap"),
+        (["#883488", "#1111119"], r"Invalid color .+ found in cmap"),
+    ],
+    ids=[
```
Comment (connortann) [inline tests/plots/test_force.py]: A minor suggestion, for readability you can provide the IDs in-line with the test case like so:

```python
from pytest import param

@pytest.mark.parametrize(
    "cmap, emsg",
    [
        param("coolwarm", None, id="valid-str"),
        ...
]

**Review by connortann — APPROVED** [review]
LGTM, with a few minor suggestions. Over to you to merge when happy!

**Comment by connortann** [comment]
> Removed most assert statements for enforcing logic, as it is not good practice

I agree asserts aren't good practice for cases where we expect the exception could be reached by the calling code, for example validating bad inputs. I think that's the case in the PR, so I agree with you here 👍

However, personally I think asserts can be helpful in other contexts, so I'd be wary of declaring them Bad Practice overall. Specifically for checking our internal logic, in cases where the assert statement _should_ always pass unless we have any bugs. It can make debugging issues easier and can also serve as documentation of any assumptions we have about the state of any variables.

**Comment by connortann** [comment]
@thatlittleboy it would be good to get this one finished off in time for our next release, if you have time this week to look at the comments above?

File: shap/plots/_force.py (hunk 1 above)
Comment (thatlittleboy) [reply shap/plots/_force.py]: Changed to `TypeError`. I also used `DimensionError` on the only other `Exception` call in `shap/plots/_force.py`.

File: tests/plots/test_force.py (hunk 2 above)
Comment (thatlittleboy) [reply tests/plots/test_force.py]: oh nice!, didn't know this option existed

**Comment by thatlittleboy** [comment]
@connortann Thanks for the comments, I made changes accordingly, please take a look when you get back.

**Comment by connortann** [comment]
LGTM!

**Pull request merged**